### PR TITLE
Add dynamic fonts loading (and downloading)

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
Add dynamic fonts loading (and downloading)
Please note that Android emulator sometimes has issues with connectivity and it needs to be restarted (mainly after the computer "goes to sleep")

## Changelog
DynamicFonts - download and load to memory fonts - this is **experimental** for now! (also requires using `react-native-fs`)

## Additional info
WOAUILIB-3752

### CODE REVIEW

**Comment by M-i-k-e-l** [comment]
@Inbal-Tish I've added the `DynamicFontsScreen`, not sure why but not all Android's fonts appear to be working, any idea?
* I do see them in the local folder

**Comment by Inbal-Tish** [comment]
Simulator Screenshot - iPhone 13 - 2023-09-11 at 09 43 24

**Comment by M-i-k-e-l** [comment]
@Inbal-Tish 
Please clean-install, it has been added to the `devDependencies`

**Comment by M-i-k-e-l** [comment]
> > @Inbal-Tish Please clean-install, it has been added to the `devDependencies` (it is also noted in the comments under changelog)
> 
> I did - twice. And pod-install

Maybe it's a cache issue? Did you clean it?
I'll retest it myself

**Comment by M-i-k-e-l** [comment]
> > > @Inbal-Tish Please clean-install, it has been added to the `devDependencies` (it is also noted in the comments under changelog)
> > 
> > 
> > I did - twice. And pod-install
> 
> Maybe it's a cache issue? Did you clean it? I'll retest it myself

Works for me :/

File: lib/components/DynamicFonts/PermissionsAcquirer.ios.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export default class PermissionsAcquirer {
+  public async checkPermissions() {
+    return Promise.resolve(true);
+  }
+
+  public async getPermissions() {
```
Comment (Inbal-Tish) [inline lib/components/DynamicFonts/PermissionsAcquirer.ios.ts:6]: You not actually checking anything... Does it work on a real device?

File: lib/components/DynamicFonts/FontLoader.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    if (!filePath) {
+      throw new Error('filePath is a required argument');
+    }
+
+    return new Promise((resolve, reject) => {
+      DynamicFont.loadFontFromFile({
+        name: fontName,
+        filePath
```
Comment (Inbal-Tish) [inline lib/components/DynamicFonts/FontLoader.ts:100]: on iOS this param is called 'path', not 'filePath'...

File: lib/android/src/main/java/com/wix/reactnativeuilib/dynamicfont/DynamicFontPackage.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import com.facebook.react.ReactPackage;
+import com.facebook.react.bridge.NativeModule;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.uimanager.ViewManager;
+import com.facebook.react.bridge.JavaScriptModule;
+public class DynamicFontPackage implements ReactPackage {
```
Comment (Inbal-Tish) [inline lib/android/src/main/java/com/wix/reactnativeuilib/dynamicfont/DynamicFontPackage.java:12]: missing new line here and in the edn of the file

File: lib/components/DynamicFonts/FontDownloader.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    if (!RNFS) {
+      throw new Error(`RNUILib FontDownloader requires installing "react-native-fs" dependency`);
+    }
+    this.fs = RNFS;
+  }
+
+  private getPrivateFolder() {
+    const {dynamicFontsFolder} = this.props;
+    return (
+      (Platform.OS === 'android' ? this.fs.ExternalDirectoryPath : this.fs.DocumentDirectoryPath) + dynamicFontsFolder
```
Comment (Inbal-Tish) [inline lib/components/DynamicFonts/FontDownloader.ts:41]: You don't need to add the 'dynamicFontsFolder' to the Android storage location? Does it not store on folders as in iOS?

File: lib/ios/reactnativeuilib/dynamicfont/DynamicFont.m [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+- (dispatch_queue_t)methodQueue
+{
+    return dispatch_get_main_queue();
+}
+RCT_EXPORT_MODULE()
+
+RCT_EXPORT_METHOD(
+  loadFontFromFile:(NSDictionary *)options callback:(RCTResponseSenderBlock)callback)
+{
+    NSString *path = [options valueForKey:@"path"];
```
Comment (Inbal-Tish) [inline lib/ios/reactnativeuilib/dynamicfont/DynamicFont.m:17]: Here - the parm name you extract is "path" (not "pathFile")

**Comment by Inbal-Tish** [comment]
@M-i-k-e-l Looks really great. I left you a few small comments and I'll keep trying to run it or see why it keeps asking me for the fs dependency...

File: lib/components/DynamicFonts/PermissionsAcquirer.ios.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+export default class PermissionsAcquirer {
+  public async checkPermissions() {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply lib/components/DynamicFonts/PermissionsAcquirer.ios.ts:2]: It is used in the same file, are you asking about the fact it is `public`?

File: lib/components/DynamicFonts/PermissionsAcquirer.ios.ts (hunk 1 above)
Comment (M-i-k-e-l) [reply lib/components/DynamicFonts/PermissionsAcquirer.ios.ts:6]: I only tested on a Simulator, do you expect it to work differently?

File: lib/components/DynamicFonts/FontLoader.ts (hunk 2 above)
Comment (M-i-k-e-l) [reply lib/components/DynamicFonts/FontLoader.ts:100]: I'll comment it out

File: lib/ios/reactnativeuilib/dynamicfont/DynamicFont.m (hunk 5 above)
Comment (M-i-k-e-l) [reply lib/ios/reactnativeuilib/dynamicfont/DynamicFont.m:17]: I'll comment it out

File: lib/android/src/main/java/com/wix/reactnativeuilib/dynamicfont/DynamicFontPackage.java (hunk 3 above)
Comment (M-i-k-e-l) [reply lib/android/src/main/java/com/wix/reactnativeuilib/dynamicfont/DynamicFontPackage.java:12]: Fixed

File: lib/components/DynamicFonts/FontDownloader.ts (hunk 4 above)
Comment (M-i-k-e-l) [reply lib/components/DynamicFonts/FontDownloader.ts:41]: It adds it to both, there are parentheses

File: lib/components/DynamicFonts/PermissionsAcquirer.ios.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply lib/components/DynamicFonts/PermissionsAcquirer.ios.ts:2]: What same file? On Android you're calling it from getPermissions but here it is not called at all...

File: lib/components/DynamicFonts/PermissionsAcquirer.ios.ts (hunk 1 above)
Comment (Inbal-Tish) [reply lib/components/DynamicFonts/PermissionsAcquirer.ios.ts:6]: Ok. If it works...

File: lib/ios/reactnativeuilib/dynamicfont/DynamicFont.m [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+@implementation DynamicFont
+
+- (dispatch_queue_t)methodQueue
+{
+    return dispatch_get_main_queue();
+}
+RCT_EXPORT_MODULE()
+
+// TODO: Needs to be tested
```
Comment (Inbal-Tish) [inline lib/ios/reactnativeuilib/dynamicfont/DynamicFont.m:14]:
```suggestion
// TODO: Needs to be tested (notice param name "path" here)
```

File: lib/components/DynamicFonts/PermissionsAcquirer.ios.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply lib/components/DynamicFonts/PermissionsAcquirer.ios.ts:2]: Oh, I think I wanted similar API, not sure why... I'll remove it

**Review by Inbal-Tish — APPROVED** [review]

**Pull request merged**